packages/protocol/test/governance/validators.ts

```diff
@@ -2205,13 +2205,116 @@ contract('Validators', (accounts: string[]) => {
+          await registerValidatorGroup(group)
+        }
+        // Start at 1 since we can't start with deaffiliate
+        for (let i = 1; i < totalEpochs; i++) {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          const blocksUntilNextEpoch = (epochNumber + 1) * EPOCH - blockNumber
+          await mineBlocks(blocksUntilNextEpoch, web3)
+
+          if (i % gapSize === 0) {
+            const group =
+              i % (gapSize * gapSize) !== 0
+                ? groups[Math.floor(i / gapSize) % groups.length]
+                : NULL_ADDRESS
+            contractIndex += 1
+            // Current epochNumber is 1 greater since we just called `mineBlocks`
+            epochs.push([epochNumber + 1, group])
+            // deaffiliate every gapSize^2 entry
+            if (i % (gapSize * gapSize) !== 0) {
+              await validators.affiliate(group)
+              await validators.addFirstMember(validator, NULL_ADDRESS, NULL_ADDRESS, {
+                from: group,
+              })
+            } else {
+              await validators.deaffiliate()
+            }
+          }
+        }
+      })
+
+      it('should correctly get the group address for exact epoch numbers', async () => {
+        for (let i = 0; i < epochs.length; i++) {
+          const group = epochs[i][1]
+          if (epochs.length - i <= membershipHistoryLength.toNumber()) {
+            assert.equal(
+              await validators.groupMembershipInEpoch(validator, epochs[i][0], 1 + i),
+              group
+            )
+          } else {
+            await assertRevert(validators.groupMembershipInEpoch(validator, epochs[i][0], 1 + i))
+          }
+        }
+      })
+
+      describe('when called with various malformed inputs', () => {
+        it('should revert when epochNumber at given index is greater than provided epochNumber', async () => {
+          await assertRevert(
+            validators.groupMembershipInEpoch(
+              validator,
+              epochs[epochs.length - 2][0],
+              contractIndex
+            )
+          )
+        })
+
+        it("should revert when epochNumber fits into a different index's bucket", async () => {
+          await assertRevert(
+            validators.groupMembershipInEpoch(
+              validator,
+              epochs[epochs.length - 1][0],
+              contractIndex - 2
+            )
+          )
+        })
+
+        it("should revert when epochNumber is greater than the chain's current epochNumber", async () => {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          await assertRevert(
+            validators.groupMembershipInEpoch(validator, epochNumber + 1, contractIndex)
+          )
+        })
+
+        it('should revert when provided index is greater than greatest index on chain', async () => {
```

Comment:
Should we check that it reverts when the provided index is less than the tail?

Reply:
Line 2273 checks this by going through all of the indices that have been "pushed out" by adding more than `membershipHistoryLength` members, but for clarity I'll add another test here